### ISSUE TITLE
Custom PCH storage location

### ISSUE THREAD

**Issue by konradsa** [clangd/clangd#255]
Is there a way to specify the pch storage location? It seems to default to /tmp. The problem is we only have a small /tmp space on our servers that are used by multiple developers. As such, it fills up quickly with pch files and then causes problems system wide. Please let me know if there is a way to change that or consider adding this option in a future version.

**Comment by ilya-biryukov** [comment]
Another workaround is to pass `-pch-storage=memory` to clangd and avoid creating those temporary files completely.
Only works if you have lots of memory to spare for clangd.

**Comment by sam-mccall** [comment]
I believe PCH storage respects any of the environment variables: `const char *EnvironmentVariables[] = {"TMPDIR", "TMP", "TEMP", "TEMPDIR"};` at least on Unix.

**Comment by ilya-biryukov** [comment]
Changing those affects other programs too, though, e.g. text editors that start clangd.

So I agree a separate flag/env var makes sense that gives a more fine-grained control would be useful.

**Comment by kadircet** [comment]
yeah i suppose there might even be value in storing them inside `.clangd/pchs/` by default

**Comment by ilya-biryukov** [comment]
Hm, I can see how .clangd is more suitable in most cases, e.g. build results normally live near the source dir, they're large and there's normally enough space to hold those.

I would personally prefer temporary files to go into /tmp and I have enough free space there, but your idea does sound like a better default.

**Comment by jrobeson** [comment]
I know it can't be done this way now, but it should be either an arg to clangd (like `--tmpdir` or an environment variable like `CLANGD_TMPDIR`. It should also be a setting in `.clangd`.  

Clangd already respects OS specific locations for its config file, so it could also use something like `$XDG_CACHE_DIR/clangd` on linux for example. This could be the default location.

**Comment by HighCommander4** [comment]
See also discussion in https://github.com/llvm/llvm-project/issues/65029
